src/formats/pnm.c

```diff
@@ -0,0 +1,287 @@
+                    v = src[x];
+                else
+                    v = src[x] << 8 | src[x + 1];
+                if (v > maxval)
+                    return PNM_EOVF;
+                if (maxval != UINT8_MAX)
+                    v = div_near(v * UINT8_MAX, maxval);
+                pix |= ARGB_SET_R(v) | ARGB_SET_G(v) | ARGB_SET_B(v);
+            } else {
+                int r, g, b;
+                if (bpc == 1) {
+                    r = src[x * 3];
+                    g = src[x * 3 + 1];
+                    b = src[x * 3 + 2];
+                } else {
+                    r = src[x * 3] << 8 | src[x * 3 + 1];
+                    g = src[x * 3 + 2] << 8 | src[x * 3 + 3];
+                    b = src[x * 3 + 4] << 8 | src[x * 3 + 5];
+                }
+                if (r > maxval || g > maxval || b > maxval)
+                    return PNM_EOVF;
+                if (maxval != UINT8_MAX) {
+                    r = div_near(r * UINT8_MAX, maxval);
+                    g = div_near(g * UINT8_MAX, maxval);
+                    b = div_near(b * UINT8_MAX, maxval);
+                }
+                pix |= ARGB_SET_R(r) | ARGB_SET_G(g) | ARGB_SET_B(b);
+            }
+            dst[x] = pix;
+        }
+    }
+    return 0;
+}
+
+enum loader_status decode_pnm(struct image* ctx, const uint8_t* data,
+                              size_t size)
+{
+    struct pnm_iter it[1];
+    bool plain;
+    int type, width, height, maxval, ret;
+
+    if (size < 2 || data[0] != 'P') {
+        return ldr_unsupported;
+    }
+    switch (data[1]) {
+        case '1':
+            plain = true;
+            type = PNM_PBM;
+            break;
+        case '2':
+            plain = true;
+            type = PNM_PGM;
+            break;
+        case '3':
+            plain = true;
+            type = PNM_PPM;
+            break;
+        case '4':
+            plain = false;
+            type = PNM_PBM;
+            break;
+        case '5':
+            plain = false;
+            type = PNM_PGM;
+            break;
+        case '6':
+            plain = false;
+            type = PNM_PPM;
+            break;
+        default:
+            return ldr_unsupported;
+    }
+    it->pos = data + 2;
+    it->end = data + size;
+
+    width = pnm_readint(it);
+    if (width < 0) {
+        image_print_error(ctx, pnm_strerror(width));
+        return ldr_fmterror;
+    }
+    height = pnm_readint(it);
+    if (height < 0) {
+        image_print_error(ctx, pnm_strerror(height));
+        return ldr_fmterror;
+    }
+    if (type != PNM_PBM) {
+        maxval = pnm_readint(it);
+        if (maxval < 0) {
+            image_print_error(ctx, pnm_strerror(maxval));
+            return ldr_fmterror;
+        }
+        if (!maxval || maxval > UINT16_MAX) {
+            image_print_error(ctx, "invalid maxval");
+            return ldr_fmterror;
+        }
+    }
+    if (!plain) {
+        if (*it->pos != ' ' && *it->pos != '\t' && *it->pos != '\n' &&
```

Comment:
Ok, I see you don't like `isspace`, I don't mind ))
Let's reduce this line in this way:
```
const char c = *it->pos;
if (c != ' ' ...
```
And the same story in `pnm_readint`.